Fix Bugzilla Issue 24458 - Mac M3 associative array keys on std.net.curl gets overwritten

MrcSnm opened the pull request:
@schveiguy Helped discover and fix the problem

schveiguy commented on std/net/curl.d:
```diff
@@ -2449,6 +2449,9 @@ struct HTTP
                     auto fieldName = m[0].toLower();
+                    ///Fixes https://issues.dlang.org/show_bug.cgi?id=24458
+                    if(fieldName is m[0])
+                        fieldName = fieldName.dup;
```
should really be `idup` since the `headersIn` is `string[string]`.

Note that issue https://issues.dlang.org/show_bug.cgi?id=2954 allows a `const char[]` to be stored as the key for an immutable `string`, but that is a different bug.

schveiguy commented:
@MrcSnm change the git commit to reference the bug to get the bot to recognize the link.

CyberShadow commented:
```d
@safe:

string[string] aa;

void f(scope const(char)[] s)
{
    aa[s] = "1";
}
```

This should not have compiled. Root cause is an accepts-invalid compiler bug?

schveiguy commented:
> Root cause is an accepts-invalid compiler bug?

Yes, see my bugzilla reference above.

schveiguy commented:
I'm unsure how we can add a unittest for this...

MrcSnm commented:
@CyberShadow are we waiting for something else?

dkorpel commented on std/net/curl.d:
```diff
@@ -2449,6 +2449,9 @@ struct HTTP
                     auto fieldName = m[0].toLower();
+                    ///Fixes https://issues.dlang.org/show_bug.cgi?id=24458
+                    if (fieldName is m[0])
+                        fieldName = fieldName.idup;
```
This should probably become an unconditional `idup` when the invalid Associative Array key type conversion bug gets fixed

MrcSnm replied:
Makes sense

MrcSnm replied:
@dkorpel Fixed it

CyberShadow replied:
That's a double copy in the worst (usual?) case. I guess that's fine; alternatively, it could be an `else fieldName = fieldName2.assumeUnique;` OSLT.

schveiguy replied:
What we need is a `toLower` that always copies.

MrcSnm replied:
@CyberShadow Made it and put type hints to it becomes clearer the fieldName is immutable

CyberShadow replied:
How about

```d
import std.exception : assumeUnique;

auto alwaysCopy(alias transformer, T)(T value)
{
    auto newValue = transformer(value);
    if (newValue is value)
        return value.idup;
    else
        return newValue.assumeUnique;
}

void main()
{
    import std.string : toLower;

    string s;
    s = "foo"; assert(s.alwaysCopy!toLower !is s);
    s = "Foo"; assert(s.alwaysCopy!toLower !is s);
}
```

MrcSnm replied:
Would that go inside std.exception library then? Btw I don't really understand why assumeUnique is related to exception but anyway.
Please tell me if there is something else need to be done.

dkorpel replied:
I thought it was good to begin with. If the test gets stuck again, I'll merge manually.

CyberShadow replied:
> Please tell me if there is something else need to be done.

We need to figure out what's with the required `dlang.phobos` check that's not letting this get green.

CyberShadow commented:
Not sure what's with the stuck `dlang.phobos` CI check. Maybe something that needs to be cherry-picked from `master`?

MrcSnm commented:
> Not sure what's with the stuck `dlang.phobos` CI check. Maybe something that needs to be cherry-picked from `master`?

I have no idea what to do either since there's no info on what's happening 🤔

CyberShadow commented:
@dkorpel Did you figure out the problem?

dkorpel commented:
No, there is no 'Details' link so I don't know what to do with it

MrcSnm commented:
> @dkorpel Did you figure out the problem?
@CyberShadow Btw, it might have something to do because it was a little overwhelming everything I've done here.

1. Initial target branch was master, which I changed to stable
2. As I worked with master, in the process of changing, it pinged like everyone as you can see on the reviewers.
3. I rewrote history so the bot also could get on bugzilla. The way I've done that was simply removing my branch which was the PR, then I force pushed the new branch which were finally synchronized with stable.

Maybe somewhere in those steps, it may have created a confusion in the runner work

dkorpel commented:
I noticed in PRs targeting master, the stuck test doesn't even exist, and https://github.com/dlang/phobos/pull/8963 has the same problem. It looks like something flaky you get when switching to stable.

CyberShadow commented:
> No, there is no 'Details' link so I don't know what to do with it

It's configured here: https://github.com/dlang/phobos/settings/branch_protection_rules/415119

There is no link because the service (whatever it is) did not report in on this PR at all.

Let me check the audit log.

> Maybe somewhere in those steps, it may have created a confusion in the runner work

I don't think you did anything that could have caused it.

CyberShadow commented:
> Let me check the audit log.

I don't see anything. Maybe it was set up before the log cutoff.
